net/resolve.go

```diff
@@ -7,26 +7,31 @@ import (
 )
 
 // ResolveUnspecifiedAddress expands an unspecified ip addresses (/ip4/0.0.0.0, /ip6/::) to
-// use the known local interfaces. If ifaceAddr is nil, we request interface addresses
-// from the network stack. (this is so you can provide a cached value if resolving many addrs)
+// use the known local interfaces
 func ResolveUnspecifiedAddress(resolve ma.Multiaddr, ifaceAddrs []ma.Multiaddr) ([]ma.Multiaddr, error) {
 	// split address into its components
-	split := ma.Split(resolve)
+	first, rest := ma.SplitFirst(resolve)
+	if first == nil {
+		return []ma.Multiaddr{resolve}, nil
+	}
 
 	// if first component (ip) is not unspecified, use it as is.
-	if !IsIPUnspecified(split[0]) {
+	if !IsIPUnspecified(first) {
 		return []ma.Multiaddr{resolve}, nil
 	}
 
 	out := make([]ma.Multiaddr, 0, len(ifaceAddrs))
 	for _, ia := range ifaceAddrs {
 		// must match the first protocol to be resolve.
-		if ia.Protocols()[0].Code != resolve.Protocols()[0].Code {
+		match := false
```

Comment:
This would be solved by #198.

Reply:
That's fair. It's still inefficient and it's very difficult to optimise this beyond the method in the PR without doing #198.